tsdb/store.go

```diff
@@ -478,21 +471,78 @@ func (s *Store) DeleteMeasurement(database, name string) error {
 		return influxql.ErrMeasurementNotFound(name)
 	}
 
+	seriesKeys := m.SeriesKeys()
+
+	shards := s.filterShards(func(sh *Shard) bool {
+		return sh.database == database
+	})
+
+	if err := s.walkShards(shards, func(sh *Shard) error {
+		if err := sh.DeleteMeasurement(m.Name, seriesKeys); err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
 	// Remove measurement from index.
 	db.DropMeasurement(m.Name)
 
-	// Remove underlying data.
+	return nil
+}
+
+// filterShards returns a slice of shards where fn returns true
+// for the shard.
+func (s *Store) filterShards(fn func(sh *Shard) bool) []*Shard {
+	shards := make([]*Shard, 0, len(s.shards))
```

Comment:
The caller is required to acquire a read lock on the `tsdb.Store`.  Looks like I missed two spots.  Generally, the exported functions acquire locks and the unexported ones assume the caller has acquired them which makes reusing the unexported functions easier to re-use.